File: packages/go/webhooks_test.go
package palpluss_test

import (
	"strings"
	"testing"

	palpluss "github.com/palpluss/palpluss-go"
)

func validWebhookPayload(eventType string) []byte {
	return []byte(`{
		"event": "transaction.updated",
		"event_type": "` + eventType + `",
		"transaction": {
			"id": "txn_webhook_001",
			"status": "SUCCESS",
			"amount": 500,
			"phone_number": "254712345678",
			"type": "STK",
			"currency": "KES",
			"provider": "mpesa",
			"tenant_id": "ten_001",
			"created_at": "2024-01-01T00:00:00Z",
			"updated_at": "2024-01-01T00:00:00Z"
		}
	}`)
}

func TestParseWebhookPayload_Valid(t *testing.T) {
	p, err := palpluss.ParseWebhookPayload(validWebhookPayload("transaction.success"))
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if p.Event != "transaction.updated" {
		t.Errorf("event: got %s", p.Event)
	}
	if p.EventType != "transaction.success" {
		t.Errorf("event_type: got %s", p.EventType)
	}
	if p.Transaction.ID != "txn_webhook_001" {
		t.Errorf("transaction.id: got %s", p.Transaction.ID)
	}
	if p.Transaction.Status != "SUCCESS" {
		t.Errorf("transaction.status: got %s", p.Transaction.Status)
	}
}

func TestParseWebhookPayload_AllEventTypes(t *testing.T) {
	eventTypes := []string{
		"transaction.success",
		"transaction.failed",
		"transaction.cancelled",
		"transaction.expired",
		"transaction.updated",
	}
	for _, et := range eventTypes {
		t.Run(et, func(t *testing.T) {
			p, err := palpluss.ParseWebhookPayload(validWebhookPayload(et))
			if err != nil {
				t.Fatalf("unexpected error for %s: %v", et, err)
			}
			if p.EventType != et {
				t.Errorf("event_type: got %s, want %s", p.EventType, et)
			}
		})
	}
}

func TestParseWebhookPayload_InvalidJSON(t *testing.T) {
	_, err := palpluss.ParseWebhookPayload([]byte("not-valid-json{"))
	if err == nil {
		t.Fatal("expected error for invalid JSON")
	}
	if !strings.Contains(err.Error(), "invalid JSON") {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestParseWebhookPayload_NotObject(t *testing.T) {
	_, err := palpluss.ParseWebhookPayload([]byte(`["array"]`))
	if err == nil {
		t.Fatal("expected error for non-object payload")
	}
}

func TestParseWebhookPayload_UnexpectedEvent(t *testing.T) {
	raw := []byte(`{
		"event": "payment.created",
		"event_type": "transaction.success",
		"transaction": {"id": "txn_001", "status": "SUCCESS"}
	}`)
	_, err := palpluss.ParseWebhookPayload(raw)
	if err == nil {
		t.Fatal("expected error for unexpected event")
	}
	if !strings.Contains(err.Error(), "unexpected event") {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestParseWebhookPayload_MissingEventType(t *testing.T) {
	raw := []byte(`{
		"event": "transaction.updated",
		"transaction": {"id": "txn_001", "status": "SUCCESS"}
	}`)
	_, err := palpluss.ParseWebhookPayload(raw)
	if err == nil {
		t.Fatal("expected error for missing event_type")
	}
	if !strings.Contains(err.Error(), "missing event_type") {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestParseWebhookPayload_UnknownEventType(t *testing.T) {
	raw := []byte(`{
		"event": "transaction.updated",
		"event_type": "transaction.refunded",
		"transaction": {"id": "txn_001", "status": "REVERSED"}
	}`)
	_, err := palpluss.ParseWebhookPayload(raw)
	if err == nil {
		t.Fatal("expected error for unknown event_type")
	}
	if !strings.Contains(err.Error(), "unknown event_type") {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestParseWebhookPayload_MissingTransaction(t *testing.T) {
	raw := []byte(`{
		"event": "transaction.updated",
		"event_type": "transaction.success"
	}`)
	_, err := palpluss.ParseWebhookPayload(raw)
	if err == nil {
		t.Fatal("expected error for missing transaction")
	}
	if !strings.Contains(err.Error(), "missing transaction") {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestParseWebhookPayload_MissingTransactionID(t *testing.T) {
	raw := []byte(`{
		"event": "transaction.updated",
		"event_type": "transaction.success",
		"transaction": {"status": "SUCCESS"}
	}`)
	_, err := palpluss.ParseWebhookPayload(raw)
	if err == nil {
		t.Fatal("expected error for missing transaction id")
	}
	if !strings.Contains(err.Error(), "missing id") {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestParseWebhookPayload_MissingTransactionStatus(t *testing.T) {
	raw := []byte(`{
		"event": "transaction.updated",
		"event_type": "transaction.success",
		"transaction": {"id": "txn_001"}
	}`)
	_, err := palpluss.ParseWebhookPayload(raw)
	if err == nil {
		t.Fatal("expected error for missing transaction status")
	}
	if !strings.Contains(err.Error(), "missing status") {
		t.Errorf("unexpected error: %v", err)
	}
}

File: packages/typescript/tests/webhooks.test.ts
import { parseWebhookPayload } from '../src/webhooks';

describe('parseWebhookPayload', () => {
  const validPayload = {
    event: 'transaction.updated',
    event_type: 'transaction.success',
    transaction: {
      id: 'tx-001',
      tenant_id: 'tenant-001',
      type: 'STK',
      status: 'SUCCESS',
      amount: 100,
      currency: 'KES',
      phone_number: '254712345678',
      external_reference: 'REF-001',
      provider: 'MPESA',
      provider_request_id: 'pr-001',
      provider_checkout_id: 'pc-001',
      result_code: '0',
      result_desc: 'Success',
      mpesa_receipt: 'QHK12345AB',
      created_at: '2024-01-01T00:00:00.000Z',
      updated_at: '2024-01-01T00:00:00.000Z',
    },
  };

  it('should parse a valid webhook payload', () => {
    const result = parseWebhookPayload(JSON.stringify(validPayload));
    expect(result.event).toBe('transaction.updated');
    expect(result.event_type).toBe('transaction.success');
    expect(result.transaction.id).toBe('tx-001');
    expect(result.transaction.mpesa_receipt).toBe('QHK12345AB');
  });

  it('should parse all valid event types', () => {
    const eventTypes = [
      'transaction.success',
      'transaction.failed',
      'transaction.cancelled',
      'transaction.expired',
      'transaction.updated',
    ];

    for (const eventType of eventTypes) {
      const payload = { ...validPayload, event_type: eventType };
      const result = parseWebhookPayload(JSON.stringify(payload));
      expect(result.event_type).toBe(eventType);
    }
  });

  it('should throw on invalid JSON', () => {
    expect(() => parseWebhookPayload('not-json')).toThrow('Invalid JSON');
  });

  it('should throw on non-object payload', () => {
    expect(() => parseWebhookPayload('"string"')).toThrow('must be a JSON object');
  });

  it('should throw on unexpected event', () => {
    const payload = { ...validPayload, event: 'unknown.event' };
    expect(() => parseWebhookPayload(JSON.stringify(payload))).toThrow('Unexpected webhook event');
  });

  it('should throw on missing event_type', () => {
    const payload = { event: 'transaction.updated', transaction: validPayload.transaction };
    expect(() => parseWebhookPayload(JSON.stringify(payload))).toThrow('missing event_type');
  });

  it('should throw on unknown event_type', () => {
    const payload = { ...validPayload, event_type: 'transaction.unknown' };
    expect(() => parseWebhookPayload(JSON.stringify(payload))).toThrow('Unknown event_type');
  });

  it('should throw on missing transaction', () => {
    const payload = { event: 'transaction.updated', event_type: 'transaction.success' };
    expect(() => parseWebhookPayload(JSON.stringify(payload))).toThrow('missing transaction');
  });

  it('should throw on missing transaction id', () => {
    const payload = {
      ...validPayload,
      transaction: { ...validPayload.transaction, id: undefined },
    };
    expect(() => parseWebhookPayload(JSON.stringify(payload))).toThrow('missing id');
  });

  it('should throw on missing transaction status', () => {
    const payload = {
      ...validPayload,
      transaction: { ...validPayload.transaction, status: undefined },
    };
    expect(() => parseWebhookPayload(JSON.stringify(payload))).toThrow('missing status');
  });

  it('should handle B2C webhook with null mpesa_receipt', () => {
    const b2cPayload = {
      ...validPayload,
      event_type: 'transaction.success',
      transaction: {
        ...validPayload.transaction,
        type: 'B2C',
        mpesa_receipt: null,
      },
    };
    const result = parseWebhookPayload(JSON.stringify(b2cPayload));
    expect(result.transaction.mpesa_receipt).toBeNull();
    expect(result.transaction.type).toBe('B2C');
  });
});
